cloudsync/youtube.py

```diff
@@ -0,0 +1,260 @@
+            self.s3.download_file(settings.VIDEO_S3_SUBTITLE_BUCKET, caption.s3_object_key, captionfile.name)
+            media_body = MediaFileUpload(captionfile.name, mimetype='mime/vtt', chunksize=-1, resumable=True)
+            if caption.language in youtube_captions:
+                return self.update_caption(media_body, youtube_captions[caption.language])
+            return self.insert_caption(caption, media_body, video_id)
+
+    def insert_caption(self, caption, media_body, video_id):
+        """
+        Upload a new video caption to YouTube
+
+        Args:
+            caption(VideoSubtitle): The VideoSubtitle to upload to YouTube
+            media_body(MediaFileUpload): The file containing the captions, in VTT format
+            video_id(str): The YouTube ID of the video to associate the captions with.
+
+        Returns:
+            dict: YouTube API response
+
+        """
+        request = self.client.captions().insert(
+            part="snippet",
+            body=dict(
+                snippet=dict(
+                    videoId=video_id,
+                    language=caption.language,
+                    name=caption.language_name,
+                    isDraft=False
+                )
+            ),
+            media_body=media_body
+        )
+        return resumable_upload(request)
+
+    def update_caption(self, media_body, caption_id):
+        """
+        Update an existing YouTube caption with a new file and return the JSON response.
+
+        Args:
+            media_body(MediaFileUpload): the video caption file to upload
+            caption_id(str): The YouTube ID of the caption
+
+        Returns:
+            dict: YouTube API response
+        """
+        request = self.client.captions().update(
+            part="snippet",
+            body=dict(
+                id=caption_id,
+                snippet=dict(
+                    isDraft=False
+                )
+            ),
+            media_body=media_body
+        )
+        return resumable_upload(request)
+
+    def delete_caption(self, caption_id):
+        """
+        Delete a caption from Youtube
+
+        Args:
+            caption_id(str): The ID of the YouTube caption
+
+        Returns:
+            int: 204 status code if successful
+        """
+        return self.client.captions().delete(id=caption_id).execute()
+
+    def upload_video(self, video, privacy='unlisted'):
+        """
+        Transfer the video's original video file from S3 to YouTube.
+        The YT account must be validated for videos > 15 minutes long:
+        https://www.youtube.com/verify
+
+        Args:
+            video(Video): The Video object whose original source file will be uploaded'
+            privacy(str): The privacy level to set the YouTube video to.
+
+        Returns:
+            dict: YouTube API response
+
+        """
+        videofile = video.videofile_set.filter(encoding=EncodingNames.ORIGINAL).first()
```

Comment:
If we expect `videofile` to always exist, can you use `get(encoding=EncodingNames.ORIGINAL)` instead?

Reply:
OK